### PR TITLE
Improve the nper function

### PR DESCRIPTION
**Pull request by cleicar**
Hi guys,

As suggested on issue #13 would be nice to raise a more descriptive message when the provided payment amount is lower than the interest amount.

Please, check it out. Thanks!

### CODE REVIEW

File: lib/exonio/financial.rb [hunk 1]
```diff
@@ -63,7 +63,14 @@ def ipmt(rate, per, nper, pv, fv = 0, end_or_beginning = 0)
     #
     def nper(rate, pmt, pv, fv = 0, end_or_beginning = 0)
       z = pmt * (1 + rate * end_or_beginning) / rate
-      temp = Math.log((-fv + z) / (pv + z))
+
+      y = (-fv + z) / (pv + z)
+
+      if y.negative?
```
Comment (mauro-oto) [inline lib/exonio/financial.rb]: @cleicar their `.travis.yml` config seems to imply that they support both Ruby 2.1 and Ruby 2.2, and this method was introduced in Ruby 2.3. Maybe consider changing this to `if y > 0`?

2.1 and 2.2 are both unsupported as of June 2018 so maybe they can consider dropping support for those versions.

File: spec/financial_spec.rb [hunk 2]
```diff
@@ -81,6 +81,10 @@
 
       expect(results).to eq(70.14566694692749)
     end
+
+    it 'raise an exception when the payment amount is lower than the interest amount' do
```
Comment (lubc) [inline spec/financial_spec.rb]: @cleicar small typo: `raise` -> `raises`

File: lib/exonio/financial.rb (hunk 1 above)
Comment (mauro-oto) [reply lib/exonio/financial.rb]: @cleicar oops, meant to say `if y < 0`, sorry about that.

File: spec/financial_spec.rb [hunk 3]
```diff
@@ -81,6 +81,10 @@
 
       expect(results).to eq(70.14566694692749)
     end
+
+    it 'raises an exception when the payment amount is lower than the interest amount' do
+      expect { Exonio.nper(rate, -10, pv, fv, 1) }.to raise_error(ArgumentError)
```
Comment (etagwerker) [inline spec/financial_spec.rb]: @cleicar Maybe to make it more explicit you could add the right message to the expectation? 

Something like this: 

```
expect { raise "oops" }.to raise_error(RuntimeError, "oops")
```

**Comment by cleicar** [comment]
@etagwerker thanks for your review. I changed the `nper` function to calculate the interest amount and use both values at the error message.

Please check it out. Thanks.

**Comment by cleicar** [comment]
@rizidoro Can we have your review in this PR, please? Thanks.

File: lib/exonio/financial.rb [hunk 4]
```diff
@@ -62,7 +62,14 @@ def ipmt(rate, per, nper, pv, fv = 0, end_or_beginning = 0)
     #   Exonio.nper(0.07 / 12, -150, 8000) # ==> 64.07334877066185
     #
     def nper(rate, pmt, pv, fv = 0, end_or_beginning = 0)
+      interest_amount = (-pv * rate) * 12
```
Comment (rizidoro) [inline lib/exonio/financial.rb]: You can remove the  `* 12` here. Just the `(-pv * rate)` will do the job =)

File: spec/financial_spec.rb [hunk 5]
```diff
@@ -81,6 +81,12 @@
 
       expect(results).to eq(70.14566694692749)
     end
+
+    it 'raises an exception when the payment amount is lower than the interest amount' do
+      raise_message = "The payment amount ($ 100) is too low. It should be at least $ 560.0 (interest amount)"
+
+      expect { Exonio.nper(rate, 100, -pv, fv, 1) }.to raise_error(ArgumentError, raise_message)
```
Comment (rizidoro) [inline spec/financial_spec.rb]: We should maintain the same Excel behavior: 

screen shot 2018-10-10 at 11 52 06

**Comment by cleicar** [comment]
@rizidoro Why should we remove the `* 12`? If we do that all the tests in `spec/financial_spec.rb` starts to fail.

Also I don't follow what you mean by `maintain the same Excel behavior`. Does it mean we should not raise an exception with the message?

File: spec/financial_spec.rb (hunk 5 above)
Comment (etagwerker) [reply spec/financial_spec.rb]: @rizidoro I think that @cleicar tried to generate a scenario for a set of parameters that will cause an error in Excel. For example:

screen shot 2018-10-12 at 8 43 03 am

(payment amount is lower than interest amount)

If I slightly vary the payment amount (to be higher than the monthly interest amount) it works fine: 

screen shot 2018-10-12 at 8 43 50 am

(payment amount is higher than interest amount)

**Comment by grzuy** [comment]
Thanks @cleicar!

Running into the same issue.

@rizidoro Any reason why this never made it to master branch? Any concerns about having this extra check?

Thanks in advance.

**Comment by rizidoro** [comment]
Hi @grzuy I'm no longer the maintainer of this package.

@PauloHerrera can you review this PR?

Thanks.

**Pull request closed without merging**